File: binary-search-tree-iterator/binary-search-tree-iterator.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 * Example of iterate a tree:
 * BSTIterator iterator = BSTIterator(root);
 * while (iterator.hasNext()) {
 *    TreeNode * node = iterator.next();
 *    do something for node
 */
using namespace std;

#include<iostream>
#include<stack>

class TreeNode {
public:
     int val;
     TreeNode *left, *right;
      TreeNode(int val) {
          this->val = val;
          this->left = this->right = NULL;
      }
  };

class BSTIterator {
public:
     stack<TreeNode *> s;
public:
    /*
    * @param root: The root of binary tree.
    */BSTIterator(TreeNode * root) {
        while(root != NULL){
            s.push(root);
            root = root->left;
        }
    }

    /*
     * @return: True if there has next node, or false
     */
    bool hasNext() {
        return !s.empty();
    }

    /*
     * @return: return next node
     */
    TreeNode * next() {
        if(!hasNext()){
            return NULL;
        }
        TreeNode *ou = s.top();
        TreeNode *out = ou;
        s.pop();

        //Caustion: time limit problem will be caused here;
        if(ou->right != NULL){
            ou = ou->right;
            while(ou != NULL){
                s.push(ou);
                ou = ou->left;
            }
        }
        
        return out;
    }
};

int main()
{
    cout<<"test"<<endl;
    return 0;
}

File: binary-search-tree-iterator/binary-search-tree-iterator.py

import queue,copy
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class BSTIterator:
    """
    @param: root: The root of binary tree.
    """
    def __init__(self, root):
        #public
        self._st = queue.LifoQueue()
        while root != None:
            self._st.put(root)
            root = root.left

    """
    @return: True if there has next node, or false
    """
    def hasNext(self, ):
        return not self._st.empty()

    """
    @return: return next node
    """
    def next(self, ):
        if not self.hasNext():return None

        ret = self._st.get()
        out = ret
        #jump to right child;
        #in order to find min at O(1), 
        #we need to push all left child into stack;
        #Caution: this are will cause time limit problem;
        if ret.right != None: 
            ret = ret.right
            while ret != None:
                self._st.put(ret)
                ret = ret.left
        return out

    def test(self):
        a = [1,2,3,4,5]
        b = a[2]
        c = b
        print(a,b,c)
        b = a[3]
        print(a, b, c)


if __name__ == "__main__":
    print(BSTIterator(TreeNode(-1)).next())
    print(BSTIterator(TreeNode(-1)).test())
